lib/sshca/identity.go

```diff
@@ -0,0 +1,384 @@
+ */
+
+// Package sshca specifies interfaces for SSH certificate authorities
+package sshca
+
+import (
+	"fmt"
+	"maps"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/crypto/ssh"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/api/constants"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/wrappers"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// Identity is a user identity. All identity fields map directly to an ssh certificate field.
+type Identity struct {
+	ValidAfter  uint64
+	ValidBefore uint64
+	// Username is teleport username
+	Username string
+	// Impersonator is set when a user requests certificate for another user
+	Impersonator string
+	// AllowedLogins is a list of SSH principals
+	AllowedLogins []string
+	// PermitX11Forwarding permits X11 forwarding for this cert
+	PermitX11Forwarding bool
+	// PermitAgentForwarding permits agent forwarding for this cert
+	PermitAgentForwarding bool
+	// PermitPortForwarding permits port forwarding.
+	PermitPortForwarding bool
+	// Roles is a list of roles assigned to this user
+	Roles []string
+	// RouteToCluster specifies the target cluster
+	// if present in the certificate, will be used
+	// to route the requests to
+	RouteToCluster string
+	// Traits hold claim data used to populate a role at runtime.
+	Traits wrappers.Traits
+	// ActiveRequests tracks privilege escalation requests applied during
+	// certificate construction.
+	ActiveRequests services.RequestIDs
+	// MFAVerified is the UUID of an MFA device when this Identity was
+	// confirmed immediately after an MFA check.
+	MFAVerified string
+	// PreviousIdentityExpires is the expiry time of the identity/cert that this
+	// identity/cert was derived from. It is used to determine a session's hard
+	// deadline in cases where both require_session_mfa and disconnect_expired_cert
+	// are enabled. See https://github.com/gravitational/teleport/issues/18544.
+	PreviousIdentityExpires time.Time
+	// LoginIP is an observed IP of the client on the moment of certificate creation.
+	LoginIP string
+	// PinnedIP is an IP from which client must communicate with Teleport.
+	PinnedIP string
+	// DisallowReissue flags that any attempt to request new certificates while
+	// authenticated with this cert should be denied.
+	DisallowReissue bool
+	// CertificateExtensions are user configured ssh key extensions (note: this field also
+	// ends up aggregating all *unknown* extensions during cert parsing, meaning that this
+	// can sometimes contain fields that were inserted by a newer version of teleport).
+	CertificateExtensions []*types.CertExtension
+	// Renewable indicates this certificate is renewable.
+	Renewable bool
+	// Generation counts the number of times a certificate has been renewed.
+	Generation uint64
```

Comment:
Can we say what `Generation: 1` means?
Does 1 mean we renewed it once or that it was never renewed?